### PR TITLE
SAMZA-2121: Add checkpoint offset field to IncomingMessageEnvelope

### PR DESCRIPTION
**Pull request by cameronlee314**
This is the first step for implementing https://issues.apache.org/jira/browse/SAMZA-2120.

### CODE REVIEW

**Comment by cameronlee314** [comment]
@rmatharu @mynameborat could you please take a look?

File: samza-core/src/test/java/org/apache/samza/task/TestAsyncRunLoop.java [hunk 1]
```diff
@@ -51,7 +51,7 @@
 import scala.Option;
 import scala.collection.JavaConverters;
 
-import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.*;
```
Comment (mynameborat) [inline samza-core/src/test/java/org/apache/samza/task/TestAsyncRunLoop.java]: nit: use explicit imports instead of `*`

**Comment by cameronlee314** [comment]
> Looks good.
> General question around checkpoint offset vs offset:
> Are there other systems that will benefit with this separation? To me it seems this separation arises only in the case of Kafka specifically large message scenario.

We don't currently have any other concrete use cases for this, but I was thinking that this was still a general enough abstraction that it could apply to other systems if they needed to modify offsets.

File: samza-api/src/main/java/org/apache/samza/system/IncomingMessageEnvelope.java [hunk 2]
```diff
@@ -86,7 +101,29 @@ public IncomingMessageEnvelope(SystemStreamPartition systemStreamPartition, Stri
...
+   * @param message A deserialized message received from the partition offset.
+   * @param size size of the message and key in bytes.
+   * @param eventTime the timestamp (in epochMillis) of when this event happened
+   * @param arrivalTime the timestamp (in epochMillis) of when this event arrived to (i.e., was picked-up by) Samza
+   */
+  public IncomingMessageEnvelope(SystemStreamPartition systemStreamPartition, String offset, String checkpointOffset,
+      Object key, Object message, int size, long eventTime, long arrivalTime) {
+    this.systemStreamPartition = systemStreamPartition;
+    this.offset = offset;
+    this.checkpointOffset = checkpointOffset;
```
Comment (shanthoosh) [inline samza-api/src/main/java/org/apache/samza/system/IncomingMessageEnvelope.java]: 1. Can you please share the rationale behind adding checkpoint offset to `IncomingMessageEnvelope` public API. 
2. Can you please throw some light on the other options considered here? Is it not plausible to do this making this change in public API. 

IMHO, it'll be better not to proliferate the public API with these internal details which will be hard to remove later on. Anyone who reads this in the future, would wonder why there are two different offsets in `IncomingMessageEnvelope` contract. Is it possible to maintain the mapping(association we want for safe-checkpoint) internally within the samza framework side and expose it to users.

Comment (cameronlee314) [reply samza-api/src/main/java/org/apache/samza/system/IncomingMessageEnvelope.java]: 1. Some more details and an example are described at https://issues.apache.org/jira/browse/SAMZA-2120.
2. We could make this part of the internal API by creating an `InternalIncomingMessageEnvelope` which is passed around internally and extends `IncomingMessageEnvelope`. Would that help avoid complicating the public API?

**Comment by cameronlee314** [comment]
This is no longer necessary. See SAMZA-2121 for more details.

**Pull request closed without merging**